src/app/3d/qgs3daxis.cpp

```diff
@@ -0,0 +1,643 @@
+  mAxisCamera->setProjectionType( mParentCamera->projectionType() );
+  // always ortho for now as it seem more "accurate"
+  if ( true || mParentCamera->projectionType() == Qt3DRender::QCameraLens::ProjectionType::OrthographicProjection )
+    mAxisCamera->lens()->setOrthographicProjection(
+      -mAxisViewportSize / 2.0f, mAxisViewportSize / 2.0f,
+      -mAxisViewportSize / 2.0f, mAxisViewportSize / 2.0f,
+      -100.0f, 500.0f );
+  else
+    mAxisCamera->lens()->setPerspectiveProjection(
+      mParentCamera->lens()->fieldOfView(), mParentCamera->lens()->aspectRatio(),
+      1.0f, -250.0f );
+
+  mAxisCamera->setUpVector( QVector3D( 0.0f, 0.0f, 1.0f ) );
+  mAxisCamera->setViewCenter( QVector3D( 0.0f, 0.0f, 0.0f ) );
+  // position will be set later
+
+  auto axisLayer = new Qt3DRender::QLayer;
+  axisLayer->setRecursive( true );
+  mAxisSceneEntity->addComponent( axisLayer );
+
+  auto axisLayerFilter = new Qt3DRender::QLayerFilter( axisViewport );
+  axisLayerFilter->addLayer( axisLayer );
+
+  auto axisCameraSelector = new Qt3DRender::QCameraSelector;
+  axisCameraSelector->setParent( axisLayerFilter );
+  axisCameraSelector->setCamera( mAxisCamera );
+
+  auto clearBuffers = new Qt3DRender::QClearBuffers( axisCameraSelector );
+  clearBuffers->setBuffers( Qt3DRender::QClearBuffers::DepthBuffer );
+
+  return axisViewport;
+}
+
+Qt3DRender::QViewport *Qgs3DAxis::constructLabelViewport( Qt3DCore::QEntity *parent3DScene, const QRectF &parentViewportSize )
+{
+  auto twoDViewport = new Qt3DRender::QViewport;
+  twoDViewport->setNormalizedRect( parentViewportSize );
+
+  mTwoDLabelSceneEntity = new Qt3DCore::QEntity;
+  mTwoDLabelSceneEntity->setParent( parent3DScene );
+  mTwoDLabelSceneEntity->setEnabled( true );
+
+  mTwoDLabelCamera = new Qt3DRender::QCamera;
+  mTwoDLabelCamera->setParent( mTwoDLabelSceneEntity );
+  mTwoDLabelCamera->setProjectionType( Qt3DRender::QCameraLens::ProjectionType::OrthographicProjection );
+  mTwoDLabelCamera->lens()->setOrthographicProjection(
+    -mParentWindow->width() / 2.0f, mParentWindow->width() / 2.0f,
+    -mParentWindow->height() / 2.0f, mParentWindow->height() / 2.0f,
+    -10.0f, 100.0f );
+
+  mTwoDLabelCamera->setUpVector( QVector3D( 0.0f, 0.0f, 1.0f ) );
+  mTwoDLabelCamera->setViewCenter( QVector3D( 0.0f, 0.0f, 0.0f ) );
+
+  mTwoDLabelCamera->setPosition( QVector3D( 0.0f, 0.0f, 100.0f ) );
+
+  auto twoDLayer = new Qt3DRender::QLayer;
+  twoDLayer->setRecursive( true );
+  mTwoDLabelSceneEntity->addComponent( twoDLayer );
+
+  auto twoDLayerFilter = new Qt3DRender::QLayerFilter( twoDViewport );
+  twoDLayerFilter->addLayer( twoDLayer );
+
+  auto twoDCameraSelector = new Qt3DRender::QCameraSelector;
+  twoDCameraSelector->setParent( twoDLayerFilter );
+  twoDCameraSelector->setCamera( mTwoDLabelCamera );
+
+  auto clearBuffers = new Qt3DRender::QClearBuffers( twoDCameraSelector );
+  clearBuffers->setBuffers( Qt3DRender::QClearBuffers::DepthBuffer );
+
+  return twoDViewport;
+}
```

Comment:
It is not clear to me why we're adding two viewports (one for axes, one for labels)... why isn't one viewport enough?

Reply:
Sorry I do not understand. Can you please explain what you mean by "labels facing the camera for a next feature" ? And how is it related to this piece of work?

I was just trying to say that I don't see a reason why there should be two viewports (with two separate branches) in the frame graph, it seems to me that one viewport should be enough - and probably also just one new branch in frame graph.

I would really like to avoid extra complexity in the frame graph unless really necessary...

Reply:
> Sorry I do not understand. Can you please explain what you mean by "labels facing the camera for a next feature" ? And how is it related to this piece of work?

I was talking about the billboarding for the labels of the axis. The idea, for a next feature we are working on, is to extract the billboarding viewport in a dedicated utility class to allow the management of billboard labels for any entity objects.

Reply:
Hmm but why would you need a dedicated billboarding viewport? We already support rendering of points as billboards, and it is integrated into the main rendering pipeline. Or maybe I am missing something here... Could you maybe post a screenshot of what you are trying to achieve with billboard labels?